### PR TITLE
Compatible with versions after dubbo-2.7.14

### PR DESCRIPTION
**Pull request by wuwen5**
Since version dubbo-2.7.14, the ```RpcContext.getContext().getAttachments()```  a copy map is returned,
so the put data will be lost.

https://github.com/apache/dubbo/blob/8240ea26cb00a52dd000d99e3997fe554134b9c3/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcContext.java#L557

https://github.com/apache/dubbo/blob/8240ea26cb00a52dd000d99e3997fe554134b9c3/dubbo-common/src/main/java/org/apache/dubbo/common/utils/MapUtils.java#L35

### CODE REVIEW

File: instrumentation/dubbo/src/test/java/brave/dubbo/ITTracingFilter_Consumer.java [hunk 1]
```diff
@@ -48,7 +48,7 @@
   @Before public void setup() {
     server.start();
 
-    String url = "dubbo://" + server.ip() + ":" + server.port() + "?scope=remote&generic=bean";
```
Comment (jcchavezs) [inline instrumentation/dubbo/src/test/java/brave/dubbo/ITTracingFilter_Consumer.java:51]: I wonder the reason for removing this query string.

**Comment by jcchavezs** [comment]
Quick question: will this continue working fine with `dubbo 2.7.8`?

**Comment by wuwen5** [comment]
> Quick question: will this continue working fine with `dubbo 2.7.8`?
Yes, it's fine

File: instrumentation/dubbo/src/test/java/brave/dubbo/ITTracingFilter_Consumer.java (hunk 1 above)
Comment (wuwen5) [reply instrumentation/dubbo/src/test/java/brave/dubbo/ITTracingFilter_Consumer.java:51]: It was used incorrectly, but did not take effect before 2.7.11，if  used ```generic=bean``` then the parameter type of ```GenericService#$invoke``` args[0] is ```JavaBeanDescriptor```

```java
public class ITTracingFilter_Provider extends ITTracingFilter {
  @Before public void setup() {
    server.service.setFilter("tracing");
    server.service.setGeneric("true");
    server.service.setInterface(GreeterService.class);
    server.service.setRef((method, parameterTypes, args) -> {
      //args[0] instanceof JavaBeanDescriptor
      String arg = (String) args[0];   // test fail
      if (arg.equals("bad")) throw new IllegalArgumentException("bad");
      return currentTraceContext.get() != null
          ? currentTraceContext.get().traceIdString()
          : "";
    });
```

see 
```GenericImplFilter#invoke```

**Review by shakuzen — APPROVED** [review]
Looks good to me. It looks like the `getObjectAttachments` method has been there for a while, so this should still work with older versions. The JavaDoc for `getAttachments` suggests using `getObjectAttachments` instead.

**Comment by wuwen5** [comment]
```java
@Experimental("Experiment api for supporting Object transmission")
public Map<String, Object> getObjectAttachments() {
    return attachments;
}
```

It looks like the ```getObjectAttachments``` method is unstable API, may get removed or changed in the next release, i think should be use ```RpcContext.getContext().setAttachment(keyName, value)```

```java
@Override protected void propagationField(String keyName, String value) {
    RpcContext.getContext().setAttachment(keyName, value);
}
```

**Comment by shakuzen** [comment]
Good catch @wuwen5. `setAttachment` does sound better, then.

**Comment by jcchavezs** [comment]
Are you OK with merging this @shakuzen ?

**Comment by codefromthecrypt** [comment]
fixed by #1371 or open a new issue. thanks!

**Pull request closed without merging**